### ISSUE TITLE
mtrand.pyx in mtrand.RandomState.beta() a<=0 error in Adaptive Greedy Algorithm

### ISSUE THREAD

**Issue by oibook13** [david-cortes/contextualbandits#7]
Hi there,

Contextual bandits problem is interesting and useful for many realistic problem. Thanks for the code!

When I tried to reproduce the online contextual bandits experiments with the file online_contextual_bandits.ipynb, I encountered an error as follows. I used python 2.7.15 (Anaconda 2.7) under Ubuntu 1604. 

I went over possible reasons, like numpy version, joblib version. It looks like somehow a non-positive number, i.e., self.a, is generated for np.random.beta(self.a, self.b, size=X.shape[0])

Thanks in advance.

`<contextualbandits.online.AdaptiveGreedy instance at 0x7f6888ee3200>

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-6-8dfd3772b2f1> in <module>()
     59                                              lst_actions[model],
     60                                              X, y,
---> 61                                              batch_st, batch_end)

<ipython-input-6-8dfd3772b2f1> in simulate_rounds(model, rewards, actions_hist, X_global, y_global, batch_st, batch_end)
     30 
     31     ## choosing actions for this batch
---> 32     actions_this_batch = model.predict(X_global[batch_st:batch_end, :]).astype('uint8')
     33 
     34     # keeping track of the sum of rewards received

/home/yluo/anaconda2/lib/python2.7/site-packages/contextualbandits/online.pyc in predict(self, X, exploit)
    877         """
    878         # TODO: add option to output scores
--> 879         return self._name_arms(self._predict(X, exploit))
    880 
    881     def _predict(self, X, exploit = False):

/home/yluo/anaconda2/lib/python2.7/site-packages/contextualbandits/online.pyc in _predict(self, X, exploit)
    898             # case 1: number of predictions to make would still fit within current window
    899             if remainder_window > X.shape[0]:
--> 900                 pred, pred_max = self._calc_preds(X)
    901                 self.window_cnt += X.shape[0]
    902                 self.window = np.r_[self.window, pred_max]

/home/yluo/anaconda2/lib/python2.7/site-packages/contextualbandits/online.pyc in _calc_preds(self, X)
    944 
    945     def _calc_preds(self, X):
--> 946         pred_proba = self._oracles.decision_function(X)
    947         pred_max = pred_proba.max(axis=1)
    948         pred = np.argmax(pred_proba, axis=1)

/home/yluo/anaconda2/lib/python2.7/site-packages/contextualbandits/utils.pyc in decision_function(self, X)
    624     def decision_function(self, X):
    625         preds = np.zeros((X.shape[0], self.n))
--> 626         Parallel(n_jobs=self.njobs, verbose=0, require="sharedmem")(delayed(self._decision_function_single)(choice, X, preds, 1) for choice in range(self.n))
    627         _apply_smoothing(preds, self.smooth, self.counters)
    628         return preds

/home/yluo/anaconda2/lib/python2.7/site-packages/joblib/parallel.pyc in __call__(self, iterable)
    915             # remaining jobs.
    916             self._iterating = False
--> 917             if self.dispatch_one_batch(iterator):
    918                 self._iterating = self._original_iterator is not None
    919 

/home/yluo/anaconda2/lib/python2.7/site-packages/joblib/parallel.pyc in dispatch_one_batch(self, iterator)
    757                 return False
    758             else:
--> 759                 self._dispatch(tasks)
    760                 return True
    761 

/home/yluo/anaconda2/lib/python2.7/site-packages/joblib/parallel.pyc in _dispatch(self, batch)
    714         with self._lock:
    715             job_idx = len(self._jobs)
--> 716             job = self._backend.apply_async(batch, callback=cb)
    717             # A job can complete so quickly than its callback is
    718             # called before we get here, causing self._jobs to

/home/yluo/anaconda2/lib/python2.7/site-packages/joblib/_parallel_backends.pyc in apply_async(self, func, callback)
    180     def apply_async(self, func, callback=None):
    181         """Schedule a func to be run"""
--> 182         result = ImmediateResult(func)
    183         if callback:
    184             callback(result)

/home/yluo/anaconda2/lib/python2.7/site-packages/joblib/_parallel_backends.pyc in __init__(self, batch)
    547         # Don't delay the application, to avoid keeping the input
    548         # arguments in memory
--> 549         self.results = batch()
    550 
    551     def get(self):

/home/yluo/anaconda2/lib/python2.7/site-packages/joblib/parallel.pyc in __call__(self)
    223         with parallel_backend(self._backend, n_jobs=self._n_jobs):
    224             return [func(*args, **kwargs)
--> 225                     for func, args, kwargs in self.items]
    226 
    227     def __len__(self):

/home/yluo/anaconda2/lib/python2.7/site-packages/contextualbandits/utils.pyc in _decision_function_single(self, choice, X, preds, depth)
    640             preds[:, choice] = self.algos[choice].predict_proba_robust(X)[:, 1]
    641         elif 'predict_proba' in dir(self.base):
--> 642             preds[:, choice] = self.algos[choice].predict_proba(X)[:, 1]
    643         else:
    644             if depth == 0:

/home/yluo/anaconda2/lib/python2.7/site-packages/contextualbandits/utils.pyc in predict_proba(self, X)
    313 
    314     def predict_proba(self, X):
--> 315         preds = np.random.beta(self.a, self.b, size=X.shape[0]).reshape((-1, 1))
    316         return np.c_[1 - preds, preds]
    317 

mtrand.pyx in mtrand.RandomState.beta()

ValueError: a <= 0

`

**Comment by david-cortes** [comment]
Thanks for the bug report!

I however tried different initializations and Python versions, but was unable to reproduce the error. Some questions:
* Are you using the latest version of this package? (that is: 1.7.2)
* Does it happen at the beginning (right after initialization) or after having called `fit`/`partial_fit` multiple times?
* Did you modify something in running the notebook?
* Are you running this in a regular computer/server/laptop, or is it some special architechture or cluster or something?
* Which NumPy and SciPy version are you using? In case they are old, have you tried updating them? (I tried NumPy 1.15 and SciPy 1.2)

It's weird that it would only happen with one class, as all classes in `online` set those numbers through the same piece of code in `utils`.

**Comment by oibook13** [comment]
Thanks for your time to check the problem.

I had figured it out, but I would answer your question before posting my solution.

- Are you using the latest version of this package? (that is: 1.7.2)
Yes, it is the latest version.
- Does it happen at the beginning (right after initialization) or after having called fit/partial_fit multiple times?
I directly used the code in online_contextual_bandits.ipynb. It took place in 
`lst_actions[model] = simulate_rounds(models[model], lst_rewards[model], lst_actions[model], X, y, batch_st, batch_end)`
- Did you modify something in running the notebook?
I didn't.
- Are you running this in a regular computer/server/laptop, or is it some special architechture or cluster or something?
I ran it on a workstation, the arch should be x86_64.
- Which joblib, NumPy and SciPy version are you using? In case they are old, have you tried updating them? (I tried NumPy 1.15 and SciPy 1.2)
I used joblib 0.12.5, numpy 1.15.4, scipy 1.1.0.

After debugging, it is caused by 
```
def _calculate_beta_prior(nchoices): 
    return (3.0/nchoices, 4)
```
in contextualbandits/utils.py line 62.
In original version, it was 3/nchoices, when nchoices=159, it turns to 0. Now, I changed it to 3.0/nchoices=0.018. Then, it works like a charm.

Another thing I would like to report is that when active_choice='weighted' for AdaptiveGreedy, it would call AdaptiveExplorer._crit_active(). This would lead to another exception 
```
TypeError: unbound method _crit_active() must be called with ActiveExplorer instance as first argument (got AdaptiveGreedy instance instead)
```
I copied AdaptiveExplorer._crit_active() to class AdaptiveGreedy and call 
```
pred[set_greedy] = np.argmax(
                self._crit_active(
                    X[set_greedy],
                    pred_all[set_greedy],
                    self.active_choice),
                axis = 1)
```
in contextualbandits/online.py line 959. Then, everything works well but I don't know whether it is correct or not.

I probably should open a new issue for this.

Thanks.

**Comment by david-cortes** [comment]
> Thanks for your time to check the problem.
> 
> I had figured it out, but I would still answer your questions before posting my solution.
> 
> * Are you using the latest version of this package? (that is: 1.7.2)
>   Yes, it is the latest version.
> * Does it happen at the beginning (right after initialization) or after having called fit/partial_fit multiple times?
>   I directly used the code in online_contextual_bandits.ipynb. It took place in
> 
> ```
> lst_actions[model] = simulate_rounds(models[model], lst_rewards[model], lst_actions[model], X, y, batch_st, batch_end)
> ```
> * Did you modify something in running the notebook?
>   I didn't.
> * Are you running this in a regular computer/server/laptop, or is it some special architechture or cluster or something?
>   I ran it on a workstation, the arch should be x86_64.
> * Which joblib, NumPy and SciPy version are you using? In case they are old, have you tried updating them? (I tried NumPy 1.15 and SciPy 1.2)
>   I used joblib 0.12.5, numpy 1.15.4, scipy 1.1.0.
> 
> After debugging, it is caused by
> 
> ```
> def _calculate_beta_prior(nchoices): 
>     return (3.0/nchoices, 4)
> ```
> in contextualbandits/utils.py line 62.
> In original version, it was 3/nchoices, when nchoices=159, it turns to 0. Now, I changed it to 3.0/nchoices=0.018. Then, it works like a charm.
> 
> Another thing I would like to report is that when active_choice='weighted' for AdaptiveGreedy, it would call AdaptiveExplorer._crit_active(). This would lead to another exception
> 
> ```
> TypeError: unbound method _crit_active() must be called with ActiveExplorer instance as first argument (got AdaptiveGreedy instance instead)
> ```
> I copied AdaptiveExplorer._crit_active() to class AdaptiveGreedy and call
> 
> ```
> pred[set_greedy] = np.argmax(
>                 self._crit_active(
>                     X[set_greedy],
>                     pred_all[set_greedy],
>                     self.active_choice),
>                 axis = 1)
> ```
> in contextualbandits/online.py line 959. Then, everything works well but I don't know whether it is correct or not.
> 
> I probably should open a new issue for this.
> 
> Thanks.

Thanks for the detailed report. It would be correct to just move the method like that. I'll try to make some slight adjustments to make it work under Python 2.7 (these are small details in which Python 2.7 and 3.4+ differ : int / int = float in 3, and the unbound method is valid in 3), although it oddly enough ran without errors for me in 2.7.15.

**Comment by oibook13** [comment]
As Python 2.7 will be deprecated on Jan 2021 (if I am not mistaken), it would be fine to support Python 3.x only. 

Your work "Adapting multi-armed bandits policies to contextual bandits scenarios" is interesting to me, do you plan to extend it to work with PyTorch or TensorFlow to increase its visibility?

**Comment by david-cortes** [comment]
Updated the potential integer divisions and unbound method calls (in `online` at least) - should work fine under Python 2.7 now. I don't plan to extend it to work with frameworks like tensorflow or pytorch, but if you want to use it with them, you'd just need to make them into a scikit-learn-like class with `fit` / `predict` (save for a few methods that use e.g. gradients) - I've tested it with xgboost for example and works fine too.
